### DISCUSSION TITLE
Is there an chance to create in an existing module an new requirement artifact, because dn_simple_createartifact.py is also creating an new modul, which i dont like.

### DISCUSSION THREAD

**Discussion by VoronLindau** [IBM/ELM-Python-Client#62] [General]
Dear Barny,
in case i still have an still existing DOORs NG Modul and i would like to add an new artifact, then how can i achieve it!?
Maybe you are able to provide me an slight modification of your "dn_simple_createartifact.py"?
Br

**Comment by barny** [comment]
I'm not sure why you say dn_simple_createartifact.py is also creating a new module, unless you've specified to it an artifact type that is a module?

The UI simplifies create a new artifact and inserting it into a module into one user action.

Using the APIs to create and add an artifact to a module you have to create the core artifact first, then add it (i.e. create a new binding) to the module structure. Check the example dn_simple_modulestructure.py - if you provide on the commandline the id of an existing artifact in the project/component it will add the artifact (i.e. create a new binding of the core artifact) to the module

**Comment by VoronLindau** [comment]
Dear Barny,
thanks for your advance, which i really appreshiate.

1.) we checked the  dn_simple_modulestructure.py. However the create Artifact is just an duplicate of an existing one containing same identifier. This is is typically not intended to have for an requirement manager. If a new artifact is create then we need also an new unique identifier for it.  Is there an change to modify the code accordingly that an new created artifact having an unique identifier!?

2.) Regarding dn_simple_createartifact.py
We called this script by given arguments:
"Requirement HelloWorld Specification_Modules"
But this script is only willing to create an artifact type "Requirement" inside an existing FolderStruct , which is typically not intended to have. Typically we would like to add an new artifact inside an existing modul. Here we make several trials , but the script is not willing in example to an in existing module, here called Artifact Type "Requirements Document" ConsumerRating, to extend it by an new artifact. When we tried to modified the start arguments to "Requirement HelloWorld ConsumerRatings" we got given failure:
**Exception: Folder 'ConsumerRatings' not found!**
Any further ideas how to achieve our expectation!?

Br Holger

![image](https://github.com/IBM/ELM-Python-Client/assets/168957149/b385bedc-e76e-4be8-b449-708df7d79371)

**Reply by barny** [reply]
The way DOORS Next works is there are core or base artifacts which a) are in a folder and b) aren't in a module, then when you want an artifact in a module you add one of these base artifacts to the module, which creates what we call a "binding". This has all the attributes of the base artifact. There are some differences between a base artifact and a binding, for example you can have links to/from a binding and to/from a base artifact. Also each binding always has a unique URL which is also different from the base artifact URL.

The UI elides much of this: it lets you (appear to) create an artifact in a module, but in the background DN is working as I described above: a base artifact is created in a folder and a binding to it is created in the module.The folder it's created in can be controlled in the Component Properties on the Module Options tab - this lets you select between `Same folder as the module` and `Specified folder` - I believe the "specified folder" is the default because when working primarily with modules this removes clutter of the core artifacts from the folder view.

> duplicate of an existing one containing same

Yes that is how DOORS Next works. If you want to create a new artifact and add it to the module, at the API level those are two separate operations: you create the base artifact in a folder, and then you create a binding for that artifact in the module.

> Any further ideas how to achieve our expectation!?

I guess your expectation has to change in the detail; you *will* create base artifacts in a folder somewhere and you *will* create bindings for these base artifacts in modules.

**Comment by VoronLindau** [comment]
Dear Barny,
we got it work.

We created first the Artifact in the Admin Area by using dn_simple_createartifact.
Afterwards binding it to the intended Modul area by using SimpleModulStructure.

Thanks for providing me/us this missing piece of this puzzle.

Br from Germany
